tuf/client_rework/metadata_bundle.py

```diff
@@ -0,0 +1,482 @@
+and deleting invalid local metadata.
+
+Loaded metadata can be accessed via the index access with rolename as key
+or, in the case of top-level metadata using the helper properties like
+'MetadataBundle.root'
+
+The rules for top-level metadata are
+ * Metadata is loadable only if metadata it depends on is loaded
+ * Metadata is immutable if any metadata depending on it has been loaded
+ * Caller must load/update these in order:
+   root -> timestamp -> snapshot -> targets -> (other delegated targets)
+ * Caller should try loading local file before updating metadata from remote
+
+Exceptions are raised if metadata fails to load in any way. The exception
+to this is local loads -- only local root metadata needs to be valid:
+other local metadata is allowed to be invalid (e.g. no longer signed):
+it won't be loaded but there will not be an exception.
+
+Example (with hypothetical download function):
+
+>>> # Load local root
+>>> bundle = MetadataBundle("path/to/metadata")
+>>>
+>>> # update root until no more are available from remote
+>>> with download("root", bundle.root.signed.version + 1) as f:
+>>>     bundle.update_root(f.read())
+>>> # ...
+>>> bundle.root_update_finished()
+>>>
+>>> # load timestamp, then update from remote
+>>> bundle.load_local_timestamp()
+>>> with download("timestamp") as f:
+>>>     bundle.update_timestamp(f.read())
+>>>
+>>> # load snapshot, update from remote if needed
+>>> if not bundle.load_local_snapshot():
+>>>     # TODO get version from timestamp
+>>>     with download("snapshot", version) as f:
+>>>         bundle.update_snapshot(f.read())
+>>>
+>>> # load local targets, update from remote if needed
+>>> if not bundle.load_local_targets():
+>>>     # TODO get version from snapshot
+>>>     with download("targets", version) as f:
+>>>         bundle.update_targets(f.read())
+>>>
+>>> # load local delegated role, update from remote if needed
+>>> if not bundle.load_local_delegated_targets("rolename", "targets"):
+>>>     # TODO get version from snapshot
+>>>     with download("rolename", version) as f:
+>>>         bundle.update_targets(f.read(), "rolename", "targets")
+
+
+TODO:
+ * exceptions are all over the place: the idea is that client could just handle a
+   generic RepositoryError that covers every issue that server provided metadata
+   could inflict (other errors would be user errors), but this is not yet the case
+ * usefulness of root_update_finished() can be debated: it could be done
+   in the beginning of load_timestamp()...
+ * there are some divergences from spec:
+   * 5.3.11: timestamp and snapshot are not deleted right away (only on next load):
+     the load functions will refuse to load the files when they are not signed by
+     current root keys. Deleting at the specified point is possible but means additional
+     code with some quirks..
+   * in general local metadata files are not deleted (they just won't succesfully load)
```

Comment:
This might be worth a discussion. I don't like deleting local metadata files for two reasons:
* it doesn't provide security (if the metadata is invalid it cannot be loaded -- if it could then our program has a bug that should be fixed: deleting the metadata won't really help)
* it prevents all debugging: "oh the app says update failed because of invalid metadata... well, the file doesn't exist :shrug:"

Reply:
Wondering is there a possible situation where you can fill up the storage by providing let's say thousands of invalid metadata files? 

Other than that your arguments are sensible.